File: app/src/main/java/com/glsebastiany/bakingapp/view/recyclerview/RVBaseAdapter.java
package com.glsebastiany.bakingapp.view.recyclerview;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;

public abstract class RVBaseAdapter extends RecyclerView.Adapter<RVGenericViewHolder> {

    @Override
    public RVGenericViewHolder onCreateViewHolder(android.view.ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = android.view.LayoutInflater.from(parent.getContext());

        android.databinding.ViewDataBinding binding = android.databinding.DataBindingUtil.inflate(
                layoutInflater, viewType, parent, false);

        return new RVGenericViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(RVGenericViewHolder holder, int position) {
        Object model = getModelObjForPosition(holder.getBinding().getRoot().getContext(), position);
        holder.bindModel(model);

        Object eventHandler = getEventHandlerObjForPosition(position);
        if (eventHandler != null) {
            holder.bindEventHandler(eventHandler);
        }
    }

    @Override
    public int getItemViewType(int position) {
        return getLayoutIdForPosition(position);
    }

    protected abstract Object getModelObjForPosition(Context context, int position);

    protected Object getEventHandlerObjForPosition(int position) {
        return null;
    }

    protected abstract int getLayoutIdForPosition(int position);
}

File: app/src/main/java/com/glsebastiany/bakingapp/view/recyclerview/RVBaseAdapter.kt

package com.glsebastiany.bakingapp.view.recyclerview

import android.support.v7.widget.RecyclerView

abstract class RVBaseAdapter : RecyclerView.Adapter<RVGenericViewHolder>() {

    override fun onCreateViewHolder(parent: android.view.ViewGroup, viewType: Int): RVGenericViewHolder {
        val layoutInflater = android.view.LayoutInflater.from(parent.context)

        val binding = android.databinding.DataBindingUtil.inflate<android.databinding.ViewDataBinding>(
                layoutInflater, viewType, parent, false)

        return RVGenericViewHolder(binding)
    }

    override fun onBindViewHolder(holder: RVGenericViewHolder,
                                  position: Int) {
        val model = getModelObjForPosition(holder.binding.root.context, position)
        holder.bindModel(model)

        val eventHandler = getEventHandlerObjForPosition(position)
        if (eventHandler != null) {
            holder.bindEventHandler(eventHandler)
        }
    }

    override fun getItemViewType(position: Int): Int = getLayoutIdForPosition(position)

    protected abstract fun getModelObjForPosition(context: android.content.Context, position: Int): Any

    open protected fun getEventHandlerObjForPosition(position: Int): Any? = null

    protected abstract fun getLayoutIdForPosition(position: Int): Int
}